### ISSUE TITLE
Am I using ChemProp Correctly?

### ISSUE THREAD

**Issue by lilleswing** [chemprop/chemprop#147]
Hi am I using the package correctly?  I am getting good results on the larger public datasets but worse results on other datasets using this method.

```
export NUM_MODELS=20
chemprop_hyperopt --data_path=time_train.csv --dataset_type=regression \
                  --target_columns=value --quiet \
                  --num_folds=5 \
                  --num_iters="$NUM_MODELS" --config_save_path=time_config.json
chemprop_train --data_path=time_train.csv --dataset_type=regression \
                --save_dir=time_model --separate_test_path=time_test.csv \
                --target_columns=value --ensemble_size=15  --config_path=time_config.json \
                --quiet
chemprop_predict --test_path=time_test.csv --checkpoint_dir=time_model --preds_path=time_preds.csv
```

Are there other flags I need to be flipping here to get better performance?

**Comment by cjmcgill** [comment]
Hi @lilleswing. There is nothing clearly wrong about what you've shown there.

There still may be room for improvement in your settings. The hyperparameter optimization using hyperopt is more directed than a random search of the space but it's still _very_ random. You may need a larger `--num_iters` than 20 in order to find a good parameter set for you.

The hyperoptimization also does not cover all the possible parameters. You may want to experiment with some others. Particularly, I would try out increasing `--epochs` to a value larger than 30. I would also try using the `--aggregation` setting for "norm" or "sum" as possible improvements as well.

There may be other things in your data handling that can help improve your performance as well. If your data spans many orders of magnitude while staying positive, a log transform of your data may make it more easily mappable for a model. If there are some atom types or functional groups that only appear a few times in your data, you may not have enough information to fit those features and they will degrade the model quality for your other predictions.

Also, Chemprop can only infer relationships between your targets and the molecular graphs. If you think that there are other inputs besides the molecular graphs that are important for the whole molecule (solvent, concentration, data source) or for individual bonds or atoms (specific bond lengths, isotope labels), there are ways to add those as well. If your molecular graphs have weird characteristics (lots of separations in the graphs) that might cause some weird model behavior as well.

It is hard to say ahead of time how much fine-tuning the model can do for you, whether you have enough data or whether the model properties are sufficiently related to the molecular graph. But these are some things that may be worth trying.

**Comment by lilleswing** [comment]
Thanks @cjmcgill  I'll add the flags to extend `--epochs` and `--features_generator rdkit_2d_normalized`.  I'm looking at automated protocols to run over the https://tdcommons.ai/single_pred_tasks/adme/#clearance-astrazeneca datasets.  20 models was picked because that can run in ~2 hrs on a 1080 TI for a 750 compound dataset.

I'll extend the training time to ~4 hours per dataset with those additional flags and see how ChemProp Does (~440 GPU hours total for the 22 datasets * 5 random seeds)!

**Comment by Evert-Homan** [comment]
So, from the answer above and your ChemRxiv paper I composed the following hyperparameter optimization settings for a "small" data set (<10K rows):

chemprop_hyperopt --data_path input.csv --dataset_type regression --features_generator rdkit_2d_normalized --no_features_scaling --aggregation sum --num_iters 100 --epochs 200 --search_parameter_keywords all
--config_save_path output_hyperopt.json

Grateful if you could comment on these settings.

Thanks for making this great tool available/Evert

**Comment by kevingreenman** [comment]
That looks good to me!

**Comment by P4j45** [comment]
Much appreciated. I am running this now on a data set just below 10K rows using an RTX5000 gpu, so will be interesting to see how long it takes.

As for training final models using optimized hyperparameters, what I gather from the documentation and paper a split is always necessary? It is not possible to use the entire data set as training set? This in order to get better coverage of features present among the training molecules.

Thanks/Evert

**Comment by kevingreenman** [comment]
The `input.csv` you give to the hyperparameter optimization should not contain any of the data that will eventually be used in your test set.

It doesn't make sense to use the entire dataset as the training set, because this then makes it impossible for you to test whether your model learned something without overfitting. It is often advisable to use a non-random splitting approach to assess the generalizability of your model to molecules it hasn't seen before.

**Comment by P4j45** [comment]
I will use the models for predicting unseen ligand only.

As for splitting the data for a final model, the resulting model will rely heavily on the selected training set if a single split is done.  Better to do x-fold CV and use those models as ensemble for external predictions? This will also assess predictive power.

**Comment by kevingreenman** [comment]
Yes, cross-validation can also be a good method for assessing the model (by looking at the average of the metric across each fold). You may also want to have a test set that is completely held out of cross validation (never in the training set) to see how the model performs on this data.

**Comment by P4j45** [comment]
Cheers. Really appreciate what you folks are doing. Best wishes/Evert
